rtrlib/rtr_mgr.c

```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
+		for (unsigned int j = 0; j < group_node->group->sockets_len;
+		     j++) {
+			rtr_stop(group_node->group->sockets[j]);
+		}
+	node = node->next;
+	}
+	pthread_mutex_unlock(&config->mutex);
+}
+
+int rtr_mgr_add_group(struct rtr_mgr_config *config,
+		      const struct rtr_mgr_group *group)
+{
+	unsigned int refresh_iv = 3600;
+	unsigned int retry_iv = 600;
+	unsigned int expire_iv = 7200;
+
+	pthread_mutex_lock(&config->mutex);
+
+	struct rtr_mgr_group_node *gnode;
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		gnode = node->data;
+		if (gnode->group->preference == group->preference) {
+			MGR_DBG1("Group with preference value already exists!");
+			pthread_mutex_unlock(&config->mutex);
+			return RTR_INVALID_PARAM;
+		}
+
+		//TODO This is not pretty. It wants to get the same intervals
+		//that are being used by other groups. Maybe intervals should
+		//be store globally/per-group/per-socket?
+		if (gnode->group->sockets[0]->refresh_interval)
+			refresh_iv = gnode->group->sockets[0]->refresh_interval;
+		if (gnode->group->sockets[0]->retry_interval)
+			retry_iv = gnode->group->sockets[0]->retry_interval;
+		if (gnode->group->sockets[0]->expire_interval)
+			expire_iv = gnode->group->sockets[0]->expire_interval;
+		node = node->next;
 	}
+	struct rtr_mgr_group *new_group = malloc(sizeof(struct rtr_mgr_group));
+
+	if (!new_group) {
+		pthread_mutex_unlock(&config->mutex);
+		return RTR_ERROR;
+	}
+	memcpy(new_group, group, sizeof(struct rtr_mgr_group));
+	new_group->status = RTR_MGR_CLOSED;
+
+	for (unsigned int j = 0; j < new_group->sockets_len; j++) {
+		if (rtr_init(new_group->sockets[j], NULL, config->pfx_table,
+			     config->spki_table, refresh_iv,
+			     expire_iv, retry_iv, rtr_mgr_cb,
+			     config, new_group) != RTR_SUCCESS) {
+			free(new_group);
+			pthread_mutex_unlock(&config->mutex);
+			return RTR_INVALID_PARAM;
+		}
+	}
+
+	struct rtr_mgr_group_node *new_group_node = malloc(
+					sizeof(struct rtr_mgr_group_node));
+	if (!new_group_node) {
+		pthread_mutex_unlock(&config->mutex);
+		free(new_group);
+		return RTR_ERROR;
+	}
+
+	new_group_node->group = new_group;
+	tommy_list_insert_tail(&config->groups, &new_group_node->node,
+			       new_group_node);
+	config->len++;
+
+	MGR_DBG("Group with preference %d successfully added!",
+		new_group->preference);
+
+	tommy_list_sort(&config->groups, &rtr_mgr_config_cmp);
+
+	struct rtr_mgr_group *best_group = rtr_mgr_get_first_group(config);
+
+	if (best_group->status == RTR_MGR_CLOSED)
+		rtr_mgr_start_sockets(best_group);
+
+	pthread_mutex_unlock(&config->mutex);
+	return RTR_SUCCESS;
+}
+
+int rtr_mgr_remove_group(struct rtr_mgr_config *config,
+			 unsigned int preference)
+{
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *remove_node = NULL;
+	tommy_node *node = tommy_list_head(&config->groups);
+	struct rtr_mgr_group_node *group_node;
+	struct rtr_mgr_group *remove_group;
+
+	// Find the node of the group we want to remove
```

Comment:
could be a separate function _find_node() or _remove_node() (which also finds it)